ledger/ledger-api-bench-tool/src/test/suite/scala/com/daml/ledger/api/benchtool/submission/CommandSubmitterITSpec.scala

```diff
@@ -0,0 +1,259 @@
+import com.daml.ledger.api.v1.transaction_service.GetTransactionTreesResponse
+import com.daml.platform.sandbox.fixture.SandboxFixture
+import org.scalactic.TripleEqualsSupport
+import org.scalatest.AppendedClues
+import org.scalatest.flatspec.AsyncFlatSpec
+import org.scalatest.matchers.should.Matchers
+import org.slf4j.{Logger, LoggerFactory}
+
+import scala.concurrent.Future
+
+object EventsObserver {
+  def apply(expectedTemplateNames: Set[String]): EventsObserver = new EventsObserver(
+    logger = LoggerFactory.getLogger(getClass),
+    expectedTemplateNames = expectedTemplateNames,
+  )
+
+  case class ObservedExerciseEvent(
+      templateName: String,
+      choiceName: String,
+      choiceArgumentsSerializedSize: Int,
+      consuming: Boolean,
+  )
+
+  case class ObservedCreateEvent(templateName: String, createArgumentsSerializedSize: Int)
+
+  case class ObservedEvents(
+      expectedTemplateNames: Set[String],
+      createEvents: Seq[ObservedCreateEvent],
+      exerciseEvents: Seq[ObservedExerciseEvent],
+  ) {
+    private val _actualTemplateNames =
+      (createEvents.map(_.templateName) ++ exerciseEvents.map(_.templateName)).toSet
+    require(
+      _actualTemplateNames.subsetOf(expectedTemplateNames),
+      s"${_actualTemplateNames} must be a subset of $expectedTemplateNames",
+    )
+
+    val consumingExercises: Seq[ObservedExerciseEvent] = exerciseEvents.filter(_.consuming)
+    val nonconsumingExercises: Seq[ObservedExerciseEvent] = exerciseEvents.filterNot(_.consuming)
+
+    val avgSizeOfConsumingExercise: Int = {
+      if (consumingExercises.isEmpty) 0
+      else consumingExercises.map(_.choiceArgumentsSerializedSize).sum / consumingExercises.size
+    }
+
+    val avgSizeOfNonconsumingExercise: Int = {
+      if (nonconsumingExercises.isEmpty) 0
+      else
+        nonconsumingExercises.map(_.choiceArgumentsSerializedSize).sum / nonconsumingExercises.size
+    }
+
+    val numberOfCreatesPerTemplateName: Map[String, Int] = {
+      val groups = createEvents.groupBy(_.templateName)
+      expectedTemplateNames.map(name => name -> groups.get(name).fold(0)(_.size)).toMap
+    }
+
+    val avgSizeOfCreateEventPerTemplateName: Map[String, Int] = {
+      val groups = createEvents.groupBy(_.templateName)
+      expectedTemplateNames.map { name =>
+        val avgSize = groups
+          .get(name)
+          .fold(0)(events =>
+            if (events.isEmpty) 0 else events.map(_.createArgumentsSerializedSize).sum / events.size
+          )
+        name -> avgSize
+      }.toMap
+    }
+
+  }
+
+}
+
+/** Collects information about create and exercise events.
+  */
+class EventsObserver(expectedTemplateNames: Set[String], logger: Logger)
+    extends ObserverWithResult[GetTransactionTreesResponse, ObservedEvents](logger) {
+
+  private val createEvents = collection.mutable.ArrayBuffer[ObservedCreateEvent]()
+  private val exerciseEvents = collection.mutable.ArrayBuffer[ObservedExerciseEvent]()
```

Comment:
This is fine as long as we don't create two subscriptions, in which case the `onNext` must be made thread-safe

Reply:
Since this is a test, the concern here is low. Feel free to encode it defensively or leave it as it is

Reply:
Leaving it as is